runtime/compiler/codegen/J9TreeEvaluator.cpp

```diff
@@ -35,6 +35,224 @@
 #include "runtime/J9ValueProfiler.hpp"
 #include "env/PersistentCHTable.hpp"
 
+void J9::TreeEvaluator::VMrdWrtbarEvaluatorForFieldWatch(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   bool isStatic = node->getSymbolReference()->getSymbol()->isStatic();
+   J9Method *owningMethod = reinterpret_cast<J9Method *>(node->getOwningMethod());
+   int32_t bcIndex = node->getByteCodeInfo().getByteCodeIndex();
+   bool isWrite = node->getOpCode().isWrtBar();
+
+   // We evaluate the children here to avoid first evaluation point in ICF later on.
+   for (int32_t i = 0; i < node->getNumChildren(); i++)
+      {
+      cg->evaluate(node->getChild(i));
+      }
+   if (isStatic)
+      {
+      // The j9classReg is the grandchild of the second child and the evaluation of the second child is optimized to avoid
+      // explicitly storing the j9class into a register. Evaluate the j9class node to store the j9class into j9classReg.
+      if (isWrite && node->getSymbolReference()->isUnresolved())
+         {
+         cg->evaluate(TR::TreeEvaluator::getDirectWrtbarJ9ClassNode(node));
```

Comment:
As mentioned in the comment above, there are scenarios where the grandchild of the second child won't be available for us in a register later on. 

Here's an x86 example provided by Yi in the `integerAddEvaluator`:
https://github.com/eclipse/omr/blob/5ceecf19aba2a9cb4e9563fda7077c29b38603f6/compiler/x/codegen/BinaryEvaluator.cpp#L740-L749

We generate the memory operation directly instead of storing the child into a register. Similar type of issues could exist on Z as well.

Reply:
I'm not understanding something. Perhaps a tree example of how exactly this may happen may help me understand.

In my mind calling `cg->evaluate` on line 48 on every child of that node will force the evaluation of all it's sub-children, including the "grandchild of the second child". If the "grandchild of the second child" had a symref which I assume it must since it's presumably a memory reference it must get evaluated and we cannot skip the evaluation (that would be a bug).

So I fail to see why we need to evaluate again here. I'd like to see a counterexample.

Reply:
Premature evaluation of children (`for` loop above this code) is removed in : a871f16

Reply:
We also don't evaluate the node returned by `getDirectWrtbarJ9ClassNode` anymore: 04f545f